### PR TITLE
Fetch vSphere credentials from vSphereClusterIdentity, instead of secret directly, in case of TMC based CSI deployment.

### PR DESCRIPTION
**Pull request by akankshapanse**
### What this PR does / why we need it
In vSphereCSIConfig controller, it derives vSphere username and password from the secret of name = <cluster_name> which is created by tanzu CLI based cluster deployment, but when TMC creates the cluster, so it won't create such secret, and the vCenter credentials are added to the VSphereClusterIdentity, instead of the secret, so CSI addon controller needs to add a way to tolerate those two scenarios:
1.CC based Cluster created by tanzu cli (in which the credential secret will be created by tanzu cli)
2.CC based Cluster created by TMC (in which the credential will be stored in VSphereClusterIdentity other than secret)
while current code only satisfies the first case, so this PR adds a fix for case 2.

### Testing Done 

Tested the fix using custom addon-manager image on setup, which had issue for CSI deployment on one of the workload cluster and the CSI deployment went fine after this change.

```
kubo@mLrfpNnL0z8NH:~$ kubectl config get-contexts
CURRENT   NAME                            CLUSTER       AUTHINFO            NAMESPACE
          cc-01-admin@cc-01               cc-01         cc-01-admin
          cc-02-admin@cc-02               cc-02         cc-02-admin
          cc-04-admin@cc-04               cc-04         cc-04-admin
*         tkg-mgmt-vc-admin@tkg-mgmt-vc   tkg-mgmt-vc   tkg-mgmt-vc-admin

kubo@mLrfpNnL0z8NH:~$ kubectl get cluster
NAME    PHASE         AGE     VERSION
cc-01   Provisioned   2d      v1.24.9+vmware.1
cc-02   Provisioned   3h31m   v1.24.9+vmware.1
cc-04   Provisioned   88m     v1.24.9+vmware.1

kubo@mLrfpNnL0z8NH:~$ kubectl get vspherecluster
NAME          READY   SERVER         AGE
cc-01-8wnhs   true    10.182.12.85   2d
cc-02-fjbhz   true    10.182.12.85   3h31m
cc-04-gk2rr   true    10.182.12.85   88m

kubo@mLrfpNnL0z8NH:~$ kubectl get vsphereclusteridentity
NAME             AGE
tkg-vc-default   2d

kubo@mLrfpNnL0z8NH:~$ kubectl describe  cluster cc-02
...
Spec:
...
  Topology:
  ...
      Variables:
      ...
            Name:                 identityRef
           Value:
            Kind:  VSphereClusterIdentity
             Name:  tkg-vc-default

kubo@mLrfpNnL0z8NH:~$ kubectl config get-contexts
CURRENT   NAME                            CLUSTER       AUTHINFO            NAMESPACE
          cc-01-admin@cc-01               cc-01         cc-01-admin
*         cc-02-admin@cc-02               cc-02         cc-02-admin            <<< workload cluster where CSI package was deployed
          tkg-mgmt-vc-admin@tkg-mgmt-vc   tkg-mgmt-vc   tkg-mgmt-vc-admin
          
 kubo@mLrfpNnL0z8NH:~$ kubectl get pods -n kube-system | grep csi
vsphere-csi-controller-6788f4d659-vklvk     7/7     Running   0               3h26m
vsphere-csi-node-44ltf                      3/3     Running   2 (3h25m ago)   3h26m
vsphere-csi-node-t596h                      3/3     Running   0               3h25m

```

Addon controller manager logs :

```
I0208 07:32:33.028179       1 clusterbootstrap_controller.go:945] ClusterBootstrapController "msg"="created the Package CR vsphere-csi.tanzu.vmware.com.2.6.2+vmware.2-tkg.1 on cluster default/cc-02"
I0208 07:32:33.062882       1 clusterbootstrap_controller.go:1018] ClusterBootstrapController "msg"="skip adding TKGS data values to secret tkg-system/cc-02-vsphere-csi-data-values because default/cc-02 is not a TKGS cluster"
I0208 07:32:33.068151       1 clusterbootstrap_controller.go:957] ClusterBootstrapController "msg"="created or patched secret tkg-system/cc-02-vsphere-csi-data-values for package vsphere-csi.tanzu.vmware.com.2.6.2+vmware.2-tkg.1 on cluster default/cc-02"
I0208 07:32:33.079492       1 clusterbootstrap_controller.go:974] ClusterBootstrapController "msg"="created or patched the PackageInstall CR tkg-system/cc-02-vsphere-csi on cluster default/cc-02"
```

### Release note

```release-note
Fetch vSphere credentials from vSphereClusterIdentity, instead of secret directly, in case of TMC based deployment of vSphere CSI.

```

### CODE REVIEW

**Comment by alfredthenarwhal** [comment]
> @jeffwubj:  /test install-vc7
Commit: 2eb5124

Tests failed! Build no: 4026

File: addons/controllers/csi/vspherecsiconfig_utils.go [hunk 1]
```diff
@@ -161,8 +168,38 @@ func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValuesNonParavirtu
...
 
-	// derive vSphere username and password from the <cluster name> secret
-	clusterSecret, err := cutil.GetSecret(ctx, r.Client, cluster.Namespace, cluster.Name)
+	// derive vSphere username and password from the <cluster name> secret.
+	// In case of TMC based deployment, we do not create <cluster name> secret.
+	// Hence if <cluster name> secret is not found, check for VSphereClusterIdentity resource
+	// to fetch vSphere username and password
+	secretName := cluster.Name
+	secretNamespace := cluster.Namespace
+	for _, variable := range cluster.Spec.Topology.Variables {
```
Comment (jeffwubj) [inline addons/controllers/csi/vspherecsiconfig_utils.go:177]: when management cluster is upgraded from legacy (none-cc) mc, will it work as expected?

**Pull request closed without merging**